File: examples/cpp/logistic_regression.cpp
// Copyright © 2023 Apple Inc.

#include <chrono>
#include <cmath>
#include <iostream>

#include "mlx/mlx.h"
#include "timer.h"

/**
 * An example of logistic regression with MLX.
 */
namespace mx = mlx::core;

int main() {
  int num_features = 100;
  int num_examples = 1'000;
  int num_iters = 10'000;
  float learning_rate = 0.1;

  // True parameters
  auto w_star = mx::random::normal({num_features});

  // The input examples
  auto X = mx::random::normal({num_examples, num_features});

  // Labels
  auto y = mx::matmul(X, w_star) > 0;

  // Initialize random parameters
  mx::array w = 1e-2 * mx::random::normal({num_features});

  auto loss_fn = [&](mx::array w) {
    auto logits = mx::matmul(X, w);
    auto scale = (1.0f / num_examples);
    return scale * mx::sum(mx::logaddexp(mx::array(0.0f), logits) - y * logits);
  };

  auto grad_fn = mx::grad(loss_fn);

  auto tic = timer::time();
  for (int it = 0; it < num_iters; ++it) {
    auto grads = grad_fn(w);
    w = w - learning_rate * grads;
    mx::eval(w);
  }
  auto toc = timer::time();

  auto loss = loss_fn(w);
  auto acc = mx::sum((mx::matmul(X, w) > 0) == y) / num_examples;
  auto throughput = num_iters / timer::seconds(toc - tic);
  std::cout << "Loss " << loss << ", Accuracy, " << acc << ", Throughput "
            << throughput << " (it/s)." << std::endl;
}

File: examples/python/logistic_regression.py
# Copyright © 2023 Apple Inc.

import time

import mlx.core as mx

lr = 0.1
num_examples = 1_000
num_features = 100
num_iters = 10_000
tic = time.time()
toc = time.time()


def loss_fn(w):
    logits = X @ w
    return mx.mean(mx.logaddexp(0.0, logits) - y * logits)


grad_fn = mx.grad(loss_fn)
# Input examples
X = mx.random.normal((num_examples, num_features))
acc = mx.mean(final_preds == y)
final_preds = (X @ w) > 0
loss = loss_fn(w)
throughput = num_iters / (toc - tic)
# Initialize random parameters
w = 1e-2 * mx.random.normal((num_features,))
# True parameters
w_star = mx.random.normal((num_features,))
# Labels
y = (X @ w_star) > 0
for _ in range(num_iters):
    grad = grad_fn(w)
    w = w - lr * grad
    mx.eval(w)
print(
    f"Loss {loss.item():.5f}, Accuracy {acc.item():.5f} "
    f"Throughput {throughput:.5f} (it/s)"
)
